### PR TITLE
Add MasterVertScale functionality to golang cluster-loader.

### PR DESCRIPTION
**Pull request by mrsiano**
This patch let the old svt MasterVertScale functionality to run on the golang cluster-loader.

### CODE REVIEW

**Comment by mrsiano** [comment]
@sjug PTAL
/cc @jeremyeder

**Comment by mrsiano** [comment]
@sjug verified and working PTAL

File: test/extended/testdata/cluster/master-vert-scale.yaml [hunk 1]
```diff
@@ -0,0 +1,65 @@
+provider: local
+ClusterLoader:
+  cleanup: true
+  projects:
+    - num: 1
```
[the remark this answers is not in the document]
Comment (jmencak) [reply test/extended/testdata/cluster/master-vert-scale.yaml]: It creates 1 project and 3 instances of build-config-template.json (line 9) within the project.  Please experiment a bit.  There is also the cluster-loader source code.

[the remark this answers is not in the document]
Comment (mrsiano) [reply test/extended/testdata/cluster/master-vert-scale.yaml]: @jmencak thanks!

**Comment by mrsiano** [comment]
@jmencak ok, so it works any other comments?! what do you think about placing the templates files under testdata ?

**Comment by jmencak** [comment]
So is this supposed to be compatible with the tests performed by the Python cluster loader?  How's IDENTIFIER passed to the templates?

**Comment by mrsiano** [comment]
@jmencak excellent point, will update the code.

File: test/extended/cluster/cl.go [hunk 2]
```diff
@@ -99,11 +100,21 @@ var _ = g.Describe("[Feature:Performance][Serial][Slow] Load cluster", func() {
...
-					config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
-					e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+					if template.Number == 0 {
+						e2e.Failf("Template %v should have at least 1 template", templateObj.Name)
+					}
+
+					identifierIdx := len(allArgs - 1) // save the identifier index
+					allArgs = append(allArgs, fmt.Sprintf("IDENTIFIER=-1"))
+					// Iterate the number of template instances
+					for template_idx := 0; template_idx < template.Number; template_idx++ {
```
Comment (mrsiano) [inline test/extended/cluster/cl.go]: @jmencak I wonder if we need to sync that template.Number to the context, specially for parallelism.
if so, I think is better to handle it in different patch ?!

File: test/extended/cluster/cl.go [hunk 3]
```diff
@@ -97,7 +97,7 @@ var _ = g.Describe("[Feature:Performance][Serial][Slow] Load cluster", func() {
 
 				// Create templates as defined
 				for _, template := range p.Templates {
-					err := CreateTemplates(oc, c, nsName, template, template.Number, tuning)
```
[the remark this answers is not in the document]
Comment (sjug) [reply test/extended/cluster/cl.go]: Yes, we can

File: test/extended/cluster/utils.go [hunk 4]
```diff
@@ -426,7 +426,8 @@ func newConfigMap(ns string, name string, vars map[string]string) *kapiv1.Config
 }
 
 // CreateTemplates creates templates in user defined namespaces with user configurable tuning sets.
-func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, template ClusterLoaderObjectType, templateCount int, tuning *TuningSetType) error {
+func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, template ClusterLoaderObjectType, tuning *TuningSetType) error {
+	//TODO: consider using a map instead of []s
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Why would we use a map?

File: test/extended/cluster/utils.go [hunk 5]
```diff
@@ -440,12 +441,24 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
 		e2e.Logf("Template environment variables will not be modified.")
 	} else {
 		params := convertVariablesToString(template.Parameters)
-		allArgs = append(allArgs, params...)
+		allArgs = append(allArgs, params)
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Why did you remove the ellipsis?

File: test/extended/cluster/utils.go [hunk 6]
```diff
@@ -440,12 +441,24 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
...
+	config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
+	e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+
+	if strings.Contains(fmt.Sprintf("%v", config), "IDENTIFIER") {
+		allArgs = append(allArgs, "--param=IDENTIFIER=-1") // tail the IDENTIFIER
+	}
+
+	for i := 0; i < template.Number; i++ {
+		// Use identifier when able
+		if strings.Contains(fmt.Sprintf("%v", config), "IDENTIFIER") {
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Why are you doing this again if you just did it outside the loop?

File: test/extended/cluster/utils.go [hunk 7]
```diff
@@ -466,6 +479,8 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
 		}
 	}
 
+
```
Comment (sjug) [inline test/extended/cluster/utils.go]: Why do we need two new lines here?

File: test/extended/cluster/utils.go (hunk 4 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: TODOs are not for casual decoration/commenting

Comment (mrsiano) [reply test/extended/cluster/utils.go]: will remove the TODO decoration.
about the map, we need to know if templates using a "identifier" parameter, if so we need to append it to args. than we iterate the number of instances and we should overwrite the identifier value, doing that by using a slice, would be not so nicer  cause we'll need to iterate the allargs to lookup the index value.
so now, I just tail the identifier if needed, and I assume it will always be the last item.
by using a map, we can have a nicer lookup. and only 1 if statement inside the loop.

File: test/extended/cluster/utils.go (hunk 5 above)
Comment (mrsiano) [reply test/extended/cluster/utils.go]: hmm by mistake, will bring it back

File: test/extended/cluster/utils.go (hunk 6 above)
Comment (mrsiano) [reply test/extended/cluster/utils.go]: to overwrite the value, else we need to rename the template name, otherwise it will fail on "already exists".
(in terms of multiple template instances).

File: test/extended/cluster/utils.go (hunk 7 above)
Comment (mrsiano) [reply test/extended/cluster/utils.go]: we don't

File: test/extended/cluster/utils.go (hunk 6 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: So why would you create it outside the loop then?

Comment (mrsiano) [reply test/extended/cluster/utils.go]: cause we use the same args but not the same identifier, so we want to overwrite just the identifier and not to append many identifiers, like:

if identifier needed -> append it to args per template
..
for loop...{
 if identifier needed -> rewrite it per instance number


doing it inside the loop will require another check, so better to have it once outside of the loop, instead multiple times in the loop

File: test/extended/cluster/utils.go (hunk 4 above)
Comment (mrsiano) [reply test/extended/cluster/utils.go]: I know, thats my was my meaning, but if we'll need to convert it to slice so as you said it might not worth it.

File: test/extended/cluster/utils.go (hunk 6 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: But you already have the identical check inside the loop already, so what are you saving?

File: test/extended/cluster/utils.go [hunk 8]
```diff
@@ -440,12 +441,24 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
...
 	}
 
-	for i := 0; i < templateCount; i++ {
-		config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
-		e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+	config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
+	e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+
+	if strings.Contains(fmt.Sprintf("%v", config), "IDENTIFIER") {
+		allArgs = append(allArgs, "--param=IDENTIFIER=-1") // tail the IDENTIFIER
```
Comment (sjug) [inline test/extended/cluster/utils.go]: This isn't really consistent with anything around parameters.

File: test/extended/cluster/utils.go [hunk 9]
```diff
@@ -440,12 +441,24 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
...
+	}
+
+	for i := 0; i < template.Number; i++ {
+		// Use identifier when able
+		if strings.Contains(fmt.Sprintf("%v", config), "IDENTIFIER") {
+			allArgs[len(allArgs)-1] = fmt.Sprintf("--param=IDENTIFIER=%v", i) // tail and reuse the IDENTIFIER instead of lookup.
+		} else {
+			// Rename the template to add another instance.
+			allArgs[0] = fmt.Sprintf("%v%v",templateObj.Name, i)
```
Comment (sjug) [inline test/extended/cluster/utils.go]: That new/renamed template doesn't exist.

**Comment by sjug** [comment]
Did you test this?

File: test/extended/cluster/utils.go (hunk 6 above)
Comment (mrsiano) [reply test/extended/cluster/utils.go]: we save the append and another check.
if we do all in the loop, it will be something like this:

for ... {
    if identifier needed {
       if first iteration do append
       else do overwrite 
      ... 
}

File: test/extended/cluster/utils.go (hunk 9 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: Where did you add it?

File: test/extended/cluster/utils.go (hunk 8 above)
Comment (sjug) [reply test/extended/cluster/utils.go]: `convertVariablesToString` is what the other agument appending function uses

File: test/extended/cluster/utils.go (hunk 9 above)
Comment (mrsiano) [reply test/extended/cluster/utils.go]: exactly by this change, line 460
allargs[0] stores the template name, it looks like previously it wasn't changed at all and "i" used only in tunings 
https://github.com/openshift/origin/blob/master/test/extended/cluster/utils.go#L446-L467

Comment (sjug) [reply test/extended/cluster/utils.go]: The new/renamed template that you're changing the name of in the argument here doesn't exist.

Comment (mrsiano) [reply test/extended/cluster/utils.go]: why do you mean ?! 
here is the first insert (index 0)
https://github.com/mrsiano/origin/blob/e245ee85b47be221b7e36b29dce7cc26a9fd7c8d/test/extended/cluster/utils.go#L438

Comment (sjug) [reply test/extended/cluster/utils.go]: That is not a template, that is an arguement in a slice.

Why do you think the `for` loop was in the location it was prior to your changes?

Comment (mrsiano) [reply test/extended/cluster/utils.go]: AFAIU, the inner for loop should create more instances for template x, originally, it uses the same template and same template name, that approach will make it failed on "already exist".

that's why I've added it, so that we can have multiple instances per template.

do you think we can make it work differently ?

Comment (sjug) [reply test/extended/cluster/utils.go]: How can you use `new-app` to create a template which does not exist?

Comment (mrsiano) [reply test/extended/cluster/utils.go]: why do you think it dose not exist?

Comment (sjug) [reply test/extended/cluster/utils.go]: Well it has to be created in order for it to exist, and it hasn't been created?

Comment (mrsiano) [reply test/extended/cluster/utils.go]: originally no, see the original code:
here we setup the args template name
https://github.com/openshift/origin/blob/86d4f945330c4161f8dd605e994c332ade6d6717/test/extended/cluster/utils.go#L437

than we iterate the template number without rename it

that works well, you can cherry pick and test it

Comment (sjug) [reply test/extended/cluster/utils.go]: I don't know how many more times I can rephrase this as you're not getting it. I am _not_ talking about the argument. You can change the template name in the arguments but that doesn't mean you have created a teamplate in OpenShift...

File: test/extended/cluster/utils.go (hunk 5 above)
Comment (jmencak) [reply test/extended/cluster/utils.go]: This needs to be fixed.  Without that, the code doesn't even compile.

Comment (mrsiano) [reply test/extended/cluster/utils.go]: +1

File: test/extended/cluster/utils.go (hunk 8 above)
Comment (jmencak) [reply test/extended/cluster/utils.go]: What will happen if IDENTIFIER is not defined by the templates?  Not every template needs to have IDENTIFIER as a paramer.  As it is, the code will break existing cluster loader functionality.

Comment (mrsiano) [reply test/extended/cluster/utils.go]: are you talking about line 450 ? if so thats why I stated the if config contains

Comment (jmencak) [reply test/extended/cluster/utils.go]: I'm talking about the fact that this PR shouldn't break existing functionality, and it does, ATM.

**Comment by mrsiano** [comment]
@jmencak @sjug PTAL the last update.
In general what changed after talk to sebastian is we want to standardise our templates to use identifier to be able to have multiple instances of that template, if the template dose not include identifier we'll raise and log message and skip it.

**Comment by jmencak** [comment]
If that's the case, OpenShift cluster loader documentation will need to updated too to reflect this change.

**Comment by mrsiano** [comment]
@ahardin-rh

File: test/extended/cluster/utils.go [hunk 10]
```diff
@@ -443,9 +443,20 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
 		allArgs = append(allArgs, params...)
 	}
 
-	for i := 0; i < templateCount; i++ {
-		config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
-		e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+	config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
+	e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+
+	if strings.Contains(fmt.Sprintf("%v", config), "IDENTIFIER") {
```
Comment (jmencak) [inline test/extended/cluster/utils.go]: This needs to be reprogrammed, full-text searching for "IDENTIFIER", really?!

Comment (mrsiano) [reply test/extended/cluster/utils.go]: Since our json's templates aren't so big, I go with this approach.
but yes we can do it nicer, we have few diffrent options:
1. split the the json from the "parameters" and scan that portion for if IDENTIFIER
2. since we gonna standardise our templates we can assume it represented and ignore this statement.
3. use templateObj.Parameters instead of template.Parameters and scan the params.

personally i prefer option number 3, which is the cheaper option, 1 requires more cpu and mem, and 2 is error prone.

could you think about other option ?

File: test/extended/cluster/utils.go [hunk 11]
```diff
@@ -443,9 +444,23 @@ func CreateTemplates(oc *exutil.CLI, c kclientset.Interface, nsName string, temp
 		allArgs = append(allArgs, params...)
 	}
 
-	for i := 0; i < templateCount; i++ {
-		config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
-		e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+	config, err := oc.AdminTemplateClient().Template().Templates(nsName).Create(templateObj)
+	e2e.Logf("Template %v created, arguments: %v, config: %+v", templateObj.Name, allArgs, config)
+
+	if strings.Contains(fmt.Sprintf("%v", templateObj.Parameters), "IDENTIFIER") {
```
Comment (mrsiano) [inline test/extended/cluster/utils.go]: @jmencak in continuing to your prv comment here we're caning only the object params

Comment (jmencak) [reply test/extended/cluster/utils.go]: Another full-text search.

Comment (mrsiano) [reply test/extended/cluster/utils.go]: this one is not full text search but, it scan only the object params and not the entire json object.

Comment (mrsiano) [reply test/extended/cluster/utils.go]: this part is pretty high performance task since it is only parse the parameters and they not so big, what we can do next is, store the params as slice and iterate the slice and return if identifier exist (than we skip other parameters if we have it).

also, we can standardise our json's to use the identifier as first one in the params, and scan only the first one.

Moreover, im gonna intorduce another PR about using stringer interfaces for types.go, which give us other capabilities in terms of parsing and abstraction.
I see the utils.go have few diffrent convert functions, which might be useful for such of similar parsing tasks.
